README.md
libTimecode
-----------

Timecode and Framerate manipulation functions.

**WORK IN PROGRESS** (Oct/31/2012)

Documentation
-------------

http://x42.github.com/libtimecode/



ChangeLog
2012-09-25 (v0.5.0) Robin Gareus <robin@gareus.org>
* refactored libltcsmpte.sf.net into libframerate


tests/tctest.c
// TODO
//
// this needs to become a proper self-test
// and example-code.
//
// currently it's a mess of things :)

#include <stdio.h>
#include <timecode/timecode.h>

int checkfps(int64_t magic, TimecodeRate const * const fps, double samplerate) {
	TimecodeTime t;
	char tcs[20];
	int64_t test;

	timecode_sample_to_time(&t, fps, samplerate, magic);
	test = timecode_to_sample(&t, fps, samplerate);
	timecode_time_to_string(&t, tcs); fprintf(stdout, "%s @%.2f%s\n", tcs, timecode_rate_to_double(fps), fps->drop?"df":"");
	printf("%lld %lld  diff: %lld\n", test, magic, magic-test);

	return 0;
}

int checkadd(TimecodeRate const * const fps, double samplerate) {
	TimecodeTime t;
	char tcs[20];
	int64_t test;

	t.hour   = 2;
	t.minute = 33;
	t.second = 59;
	t.frame  = 22;
	t.subframe = 20;

	test = timecode_to_sample(&t, fps, samplerate);
	test*=2;
	timecode_sample_to_time(&t, fps, samplerate, test);
	timecode_time_to_string(&t, tcs); fprintf(stdout, "%s\n", tcs);
	test/=2;
	timecode_sample_to_time(&t, fps, samplerate, test);

	t.hour   = 2;
	t.minute = 33;
	t.second = 59;
	t.frame  = 22;
	t.subframe = 20;
	timecode_time_add(&t, fps, &t, &t);
	timecode_time_to_string(&t, tcs); fprintf(stdout, "%s\n", tcs);
	return 0;
}

int checksub(TimecodeRate const * const fps, double samplerate) {
	TimecodeTime t1,t2,res;
	char tcs[20];
	int64_t test1,test2;

	t1.hour   = 2;
	t1.minute = 33;
	t1.second = 59;
	t1.frame  = 22;
	t1.subframe = 0;

	t2.hour   = 0;
	t2.minute = 20;
	t2.second = 9;
	t2.frame  = 12;
	t2.subframe = 50;

	test1 = timecode_to_sample(&t1, fps, samplerate);
	test2 = timecode_to_sample(&t2, fps, samplerate);

	timecode_sample_to_time(&res, fps, samplerate, test1-test2);
	timecode_time_to_string(&res, tcs); fprintf(stdout, "%s\n", tcs);

	timecode_time_subtract(&res, fps, &t1, &t2);
	timecode_time_to_string(&res, tcs); fprintf(stdout, "%s\n", tcs);
	return 0;
}

int checkcmp() {
	Timecode a,b;

	a.d.year     = 2008;
	a.d.month    = 12;
	a.d.day      = 31;
	a.d.timezone = 0;
	a.t.hour     = 23;
	a.t.minute   = 59;
	a.t.second   = 59;
	a.t.frame    = 29;
	a.t.subframe = 29;

	b.d.year     = 2008;
	b.d.month    = 12;
	b.d.day      = 31;
	b.d.timezone = -60;
	b.t.hour     = 23;
	b.t.minute   = 59;
	b.t.second   = 59;
	b.t.frame    = 29;
	b.t.subframe = 29;

	printf("compare a<>b = %d\n", 
			timecode_datetime_compare(TCFPS2997DF, &a, &b));
	return 0;
}

int main (int argc, char **argv) {
  int64_t magic = 964965602; // 05:34:43:11 @29.97ndf, 48kSPS
  //int64_t magic = 1601568888; //
	Timecode tc;

	/* test converter */
	printf("test convert\n");
	checkfps(magic, TCFPS24, 48000);
	checkfps(magic, TCFPS25, 48000);
	checkfps(magic, TCFPS2997DF, 48000);
	checkfps(magic, &tcfps2997ndf, 48000);
	checkfps(magic, TCFPS30, 48000);

	/* test parser */
	printf("test parser\n");
	TimecodeTime t;
	char tcs[20];
	timecode_parse_time(&t, TCFPS25, "1:::-1");
	timecode_time_to_string(&t, tcs); fprintf(stdout, "%s\n", tcs);

	/* test add/sub */
	printf("test addition/subtraction\n");
	checkadd(TCFPS2997DF, 48000);
	checksub(TCFPS2997DF, 48000);

	checkadd(TCFPS30, 48000);
	checksub(TCFPS30, 48000);

	/* test add/sub */
	printf("test compare\n");
	checkcmp();

	/* test inc/dec */
	printf("test inc/dec\n");
	tc.d.year     = 2008;
	tc.d.month    = 12;
	tc.d.day      = 31;
	tc.d.timezone = 0;
	tc.t.hour     = 23;
	tc.t.minute   = 59;
	tc.t.second   = 59;
	tc.t.frame    = 29;
	tc.t.subframe = 29;

	timecode_datetime_increment(&tc, TCFPS30);
	timecode_datetime_decrement(&tc, TCFPS30);

	fprintf(stdout, "%02d/%02d/%02d  %02d:%02d:%02d%c%02d\n",
			tc.d.month,
			tc.d.day,
			tc.d.year,

			tc.t.hour,
			tc.t.minute,
			tc.t.second,
			(TCFPS30->drop) ? '.' : ':',
			tc.t.frame
			);

	tc.t.subframe = 0;

	timecode_time_to_string(&tc.t, tcs); fprintf(stdout, "%s @30fps\n", tcs);
	timecode_convert_rate(&tc.t, TCFPS24, &tc.t, TCFPS30);
	timecode_time_to_string(&tc.t, tcs); fprintf(stdout, "%s @24fps\n", tcs);
	timecode_convert_rate(&tc.t, TCFPS25, &tc.t, TCFPS24);
	timecode_time_to_string(&tc.t, tcs); fprintf(stdout, "%s @25fps\n", tcs);
	timecode_convert_rate(&tc.t, TCFPS30, &tc.t, TCFPS25);
	timecode_time_to_string(&tc.t, tcs); fprintf(stdout, "%s @30fps\n", tcs);
	timecode_convert_rate(&tc.t, TCFPS25, &tc.t, TCFPS30);
	timecode_time_to_string(&tc.t, tcs); fprintf(stdout, "%s @25fps\n", tcs);
	timecode_convert_rate(&tc.t, TCFPS24, &tc.t, TCFPS25);
	timecode_time_to_string(&tc.t, tcs); fprintf(stdout, "%s @24fps\n", tcs);
	timecode_convert_rate(&tc.t, TCFPS30, &tc.t, TCFPS24);
	timecode_time_to_string(&tc.t, tcs); fprintf(stdout, "%s @30fps\n", tcs);
	timecode_convert_rate(&tc.t, &tcfpsUS, &tc.t, TCFPS30);
	timecode_time_to_string(&tc.t, tcs); fprintf(stdout, "%s\n", tcs);

	return 0;
}


release.sh
#!/bin/sh

EDITOR=${EDITOR:-editor}

if test ! -d .git; then
	echo "ERROR: This script runs only from a git repository."
	exit 1
fi

if test "`git symbolic-ref HEAD`" != "refs/heads/master"; then
	echo "ERROR: The git 'master' branch must be checked out."
	exit 1
fi

echo "commit pending changes.."
git commit -a 

# to bump the version number
# - src/timecode/timecode.h (this is the MAIN VERSION used by configure) 
#   and re-run autogen.sh to pick it up.
# - Changelog 
# - debian/changelog (for the debian package)

echo "Update version number -- edit 3 files: src/timecode/timecode.h ChangeLog debian/changelog"
echo -n "launch editor ? [Y/n]"
read -n1 a
echo

if test "$a" != "n" -a "$a" != "N"; then
	${EDITOR} src/timecode/timecode.h debian/changelog ChangeLog
	sh autogen.sh
fi

VERSION=$(awk '/ VERSION /{print $3;}' src/config.h | sed 's/"//g')

if [ -z "$VERSION" ]; then
  echo "unknown VERSION number"
  exit 1;
fi

echo "VERSION: $VERSION"

echo -n "Is this correct? [Y/n]"
read -n1 a
echo
if test "$a" == "n" -o "$a" == "N"; then
	exit 1
fi

echo "re-creating man-pages and documentation with new version-number.."
make dox 

echo "creating git-commit of updated doc & version number"
git commit -m "finalize changelog v${VERSION}" src/timecode/timecode.h ChangeLog debian/changelog doc/man/man3/timecode.h.3

cd doc/html
git add *.*
git commit -a -m "doxgen -- v${VERSION}"
cd ../..

git tag "v${VERSION}" || (echo -n "version tagging failed. - press Enter to continue, CTRL-C to stop."; read; ) 

echo -n "git push? [Y/n]"
read -n1 a
echo

if test "$a" != "n" -a "$a" != "N"; then
	git push origin || exit 1
	#git push --tags ## would push ALL existing tags,
	git push origin "refs/tags/v${VERSION}:refs/tags/v${VERSION}" || exit 1
	cd doc/html
	git push origin gh-pages || exit 1
	cd ../..
fi

make dist
ls -l "libtimecode-${VERSION}.tar.gz"

GITREPO="x42/libtimecode"
x-www-browser https://github.com/${GITREPO}/downloads


build-deb.sh
#!/bin/sh

PACKAGE=libtimecode
DEBRELEASE=$(head -n1 debian/changelog | cut -d ' ' -f 2 | sed 's/[()]*//g')

TMPDIR=/tmp/$PACKAGE-${DEBRELEASE}
rm -rf ${TMPDIR}

GITBRANCH=${GITBRANCH:-master}

echo "debian -export-ignore" >> .git/info/attributes

git-buildpackage \
	--git-upstream-branch=$GITBRANCH --git-debian-branch=$GITBRANCH \
	--git-upstream-tree=branch \
	--git-export-dir=${TMPDIR} --git-cleaner=/bin/true \
	--git-force-create \
	-rfakeroot $@ 

ERROR=$?

ed -s .git/info/attributes > /dev/null << EOF
/debian -export-ignore
d
wq
EOF

if test $ERROR != 0; then
	exit $ERROR
fi

lintian -i --pedantic ${TMPDIR}/${PACKAGE}_${DEBRELEASE}_*.changes \
	| tee /tmp/${PACKAGE}.issues

echo
ls ${TMPDIR}/${PACKAGE}_${DEBRELEASE}_*.changes
echo
echo dput rg42 ${TMPDIR}/${PACKAGE}_${DEBRELEASE}_*.changes


src/timecode.c
/*
   libtimecode - convert A/V timecode and framerate

   Copyright (C) 2006, 2007, 2008, 2012 Robin Gareus <robin@gareus.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU Lesser Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Lesser Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

*/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "timecode/timecode.h"

/*****************************************************************************
 * misc & helper functions
 */

#define TCtoDbl(r) ( (double)((r)->num) / (double)((r)->den) )

double timecode_rate_to_double(TimecodeRate const * const r) {
	return TCtoDbl(r);
}

double timecode_frames_per_timecode_frame(TimecodeRate const * const r, const double samplerate) {
	return (samplerate / TCtoDbl(r));
}

/*****************************************************************************
 * timecode <> sample,frame-number
 */

int64_t timecode_to_sample (TimecodeTime const * const t, TimecodeRate const * const r, const double samplerate) {
	const double  fps_d = TCtoDbl(r);
	const int64_t fps_i = ceil(fps_d);
	const double frames_per_timecode_frame = samplerate / fps_d;
	int64_t sample;

	if (r->drop) {
		int64_t totalMinutes = 60 * t->hour + t->minute;
		int64_t frameNumber  = fps_i * 3600 * t->hour + fps_i * 60 * t->minute
			                 + fps_i * t->second + t->frame
					 - 2 * (totalMinutes - totalMinutes / 10);

		sample = frameNumber * samplerate / fps_d;
	} else {
		sample = (int64_t) rint(
				(
				   ((t->hour * 60 * 60) + (t->minute * 60) + t->second)
				 * (fps_i * frames_per_timecode_frame)
				)
				 + (t->frame * frames_per_timecode_frame));
	}
	if (r->subframes != 0) {
		sample += t->subframe * frames_per_timecode_frame / r->subframes;
	}
	return sample;
}

void timecode_sample_to_time (TimecodeTime * const t, TimecodeRate const * const r, const double samplerate, const int64_t sample) {
	const double  fps_d = TCtoDbl(r);
	const int64_t fps_i = ceil(fps_d);

	if (r->drop) {
		int64_t frameNumber = floor( (double)sample * fps_d / samplerate );

		/* there are 17982 frames in 10 min @ 29.97df */
		const int64_t D = frameNumber / 17982;
		const int64_t M = frameNumber % 17982;

		t->subframe =  rint(r->subframes * ((double)sample * fps_d / samplerate - (double)frameNumber));

		if (t->subframe == r->subframes) {
			t->subframe = 0;
			frameNumber++;
		}

		frameNumber +=  18*D + 2*((M - 2) / 1798);

		t->frame  =    frameNumber % 30;
		t->second =   (frameNumber / 30) % 60;
		t->minute =  ((frameNumber / 30) / 60) % 60;
		t->hour   = (((frameNumber / 30) / 60) / 60);

	} else {
		double timecode_frames_left_exact;
		double timecode_frames_fraction;
		int64_t timecode_frames_left;
		const double frames_per_timecode_frame = samplerate / fps_d;
		const int64_t frames_per_hour = (int64_t)(3600 * fps_i * frames_per_timecode_frame);

		t->hour = sample / frames_per_hour;
		double sample_d = sample % frames_per_hour;

		timecode_frames_left_exact = sample_d / frames_per_timecode_frame;
		timecode_frames_fraction = timecode_frames_left_exact - floor( timecode_frames_left_exact );

		t->subframe = (int32_t) rint(timecode_frames_fraction * r->subframes);

		timecode_frames_left = (int64_t) floor (timecode_frames_left_exact);

		if (t->subframe == r->subframes) {
			t->subframe = 0;
			timecode_frames_left++;
		}

		t->minute = timecode_frames_left / (fps_i * 60);
		timecode_frames_left = timecode_frames_left % (fps_i * 60);
		t->second = timecode_frames_left / fps_i;
		t->frame  = timecode_frames_left % fps_i;
	}
}

int64_t timecode_to_framenumber (TimecodeTime const * const t, TimecodeRate const * const r) {
	return timecode_to_sample(t, r, TCtoDbl(r));
}

void timecode_framenumber_to_time (TimecodeTime * const t, TimecodeRate const * const r, const int64_t frameno) {
	timecode_sample_to_time(t, r, TCtoDbl(r), frameno);
}

void timecode_convert_rate (TimecodeTime * const t_out, TimecodeRate const * const r_out, TimecodeTime * const t_in, TimecodeRate const * const r_in) {
	//const double rate = 84672000; // LCM(192k, 88.2k, 24, 25, 30)
	const double rate = TCtoDbl(r_out) < TCtoDbl(r_in) ?  TCtoDbl(r_in) :  TCtoDbl(r_out);
	int64_t s = timecode_to_sample(t_in, r_in, rate);
	timecode_sample_to_time(t_out, r_out, rate, s);
}

/*****************************************************************************
 * Add Subtract
 */

static int32_t move_overflow(TimecodeTime * const t, TimecodeRate const * const r) {
	int i;
	int32_t rv = 0;
	int32_t * const bcd[6] = {&t->subframe, &t->frame, &t->second, &t->minute, &t->hour, &rv };

	int smpte_table[6] =  {1, 1, 60, 60, 24, 0 };
	smpte_table[0] = r->subframes;
	smpte_table[1] = ceil(TCtoDbl(r));

	for (i=0; i<5; i++) {
		if ((*bcd[i] >= smpte_table[i]) || (*bcd[i] < 0) ) {
			int ov= (int) floor((double) (*bcd[i]) / smpte_table[i]);
#if 0
			// TODO drop-frames ? - basically only needed when parsing invalid TC
			// via timecode_parse_time()
			if (r->drop && i==2) minutes_crossed += ov;
			if (r->drop && i==3) minutes_crossed += ov * 60;
			//if (r->drop && i==4) minutes_crossed += ov * 60 * 24;
#endif
			*bcd[i] -= ov*(smpte_table[i]);
			if (bcd[i+1]) { *bcd[i+1]+=ov; }
		}
	}

#if 0
	if (r->drop) {
		// 108 drop-frames per hour, but..
	}
#endif
	return rv;
}

static int32_t dropped_frames(TimecodeTime const * const t) {
		int64_t totalMinutes = 60 * t->hour + t->minute;
		return 2 * (totalMinutes - totalMinutes / 10);
}

void timecode_time_add (TimecodeTime * const res, TimecodeRate const * const r, TimecodeTime const * const t1, TimecodeTime const * const t2) {
	int df = 0;
	if (r->drop) {
		df = dropped_frames(t1) + dropped_frames(t2);
	}

	res->subframe = t1->subframe + t2->subframe;
	res->frame    = t1->frame    + t2->frame   ;
	res->second   = t1->second   + t2->second  ;
	res->minute   = t1->minute   + t2->minute  ;
	res->hour     = t1->hour     + t2->hour    ;

	if (r->drop) {
		move_overflow(res, r);
		res->frame += dropped_frames(res) - df;
	}

	move_overflow(res, r);
}

void timecode_time_subtract (TimecodeTime * const res, TimecodeRate const * const r, TimecodeTime const * const t1, TimecodeTime const * const t2) {
	int df = 0;
	if (r->drop) {
		df = dropped_frames(t1) - dropped_frames(t2);
	}

	res->subframe = t1->subframe - t2->subframe;
	res->frame    = t1->frame    - t2->frame   ;
	res->second   = t1->second   - t2->second  ;
	res->minute   = t1->minute   - t2->minute  ;
	res->hour     = t1->hour     - t2->hour    ;

	if (r->drop) {
		move_overflow(res, r);
		res->frame += dropped_frames(res) - df;
	}

	move_overflow(res, r);
}

#define CMP(a,b) ( (a) > (b) ? 1 : -1)
int timecode_time_compare (TimecodeRate const * const r, TimecodeTime const * const a, TimecodeTime const * const b) {
	if (a->hour     != b->hour    ) return CMP(a->hour, b->hour);
	if (a->minute   != b->minute  ) return CMP(a->minute, b->minute);
	if (a->second   != b->second  ) return CMP(a->second, b->second);
	if (a->frame    != b->frame   ) return CMP(a->frame, b->frame);
	if (a->subframe != b->subframe) return CMP(a->subframe, b->subframe);
	return (0);
}

int timecode_date_compare (TimecodeDate const * const a, TimecodeDate const * const b) {
	if (a->year  != b->year ) return CMP(a->year, b->year);
	if (a->month != b->month) return CMP(a->month, b->month);
	if (a->day   != b->day  ) return CMP(a->day, b->day);
	// TODO timezone  -- check intl. date-line crossing.
	if (a->timezone != b->timezone) return CMP(a->timezone, b->timezone);
	return (0);
}

int timecode_datetime_compare (TimecodeRate const * const r, Timecode const * const a, Timecode const * const b) {
	Timecode ax, bx;
	memcpy(&ax, a, sizeof(Timecode));
	memcpy(&bx, b, sizeof(Timecode));

	/* convert to UTC */
	ax.t.minute -= a->d.timezone;
	bx.t.minute -= b->d.timezone;
	ax.d.timezone = bx.d.timezone = 0;

	/* adjust day, month */
	int ao = move_overflow(&ax.t, r);
	int bo = move_overflow(&bx.t, r);

	if (ao < 0) {
		int i;
		for (i=0; i > ao; --i) {
			timecode_date_decrement(&ax.d);
		}
	} else if (ao > 0) {
		int i;
		for (i=0; i < ao; ++i) {
			timecode_date_increment(&ax.d);
		}
	}

	if (bo < 0) {
		int i;
		for (i=0; i > bo; --i) {
			timecode_date_decrement(&bx.d);
		}
	} else if (bo > 0) {
		int i;
		for (i=0; i < bo; ++i) {
			timecode_date_increment(&bx.d);
		}
	}

	int rv;
	if ((rv=timecode_date_compare(&ax.d, &bx.d)) != 0)
		return rv;

	return (timecode_time_compare(r, &ax.t, &bx.t));
}

/*****************************************************************************
 * Increment/Decrement
 */

void timecode_date_increment(TimecodeDate * const d) {
	unsigned char dpm[12] = {31,28,31,30,31,30,31,31,30,31,30,31};
	if ( (d->year%4)==0 && ( (d->year%100) != 0 || (d->year%400) == 0) ) {
		dpm[1]=29;
	}

	d->day++;
	if (d->day > dpm[d->month-1]) {
		d->day=1;
		d->month++;
		if (d->month > 12) {
			d->month = 1;
			d->year++;
		}
	}
}

int timecode_time_increment(TimecodeTime * const t, TimecodeRate const * const r) {
	int rv = 0;
	const int fps = ceil(TCtoDbl(r));
	t->frame++;

	if (fps < t->frame) goto done;
	t->frame = 0;
	t->second++;
	if (t->second < 60) goto done;
	t->second = 0;
	t->minute++;
	if (t->minute < 60) goto done;
	t->minute = 0;
	t->hour++;
	if (t->hour < 24) goto done;
	t->hour = 0;
	rv=1;

done:
	if (r->drop && (t->minute%10 != 0) && (t->second == 0) && (t->frame == 0)) {
		t->frame += 2;
	}
	return rv;
}

void timecode_date_decrement (TimecodeDate * const d) {
	unsigned char dpm[12] = {31,28,31,30,31,30,31,31,30,31,30,31};
	if ( (d->year%4)==0 && ( (d->year%100) != 0 || (d->year%400) == 0) ) {
		dpm[1]=29;
	}

	if (d->day > 1) {
		d->day--;
		return;
	}
	d->month = 1 + (d->month + 10)%12;
	d->day = dpm[d->month-1];

	if (d->month == 12) {
			d->year--;
	}
}

int timecode_time_decrement(TimecodeTime * const t, TimecodeRate const * const r) {
	const int fps = ceil(TCtoDbl(r));

	if (r->drop && (t->minute%10 != 0) && (t->second == 0) && (t->frame == 2)) {
		; // assume t->frame==0;
	} else

	if (t->frame > 0) {
		t->frame--;
		return 0;
	}

	t->frame = fps-1;
	if (t->second > 0) {
		t->second--;
		return 0;
	}

	t->second = 59;
	if (t->minute > 0) {
		t->minute--;
		return 0;
	}

	t->minute = 59;
	if (t->hour > 0) {
		t->hour--;
		return 0;
	}
	t->hour = 23;

	return 1;
}

int timecode_datetime_increment (Timecode * const dt, TimecodeRate const * const r) {
	if (timecode_time_increment(&dt->t, r)) {
		timecode_date_increment(&dt->d);
		return 1;
	}
	return 0;
}

int timecode_datetime_decrement (Timecode * const dt, TimecodeRate const * const r) {
	if (timecode_time_decrement(&dt->t, r)) {
		timecode_date_decrement(&dt->d);
		return 1;
	}
	return 0;
}

/*****************************************************************************
 * Format & Parse
 */

void timecode_datetime_to_string (Timecode const * const tc, char *smptestring) {
	snprintf(smptestring, 33, "%02d/%02d/%04d %02d:%02d:%02d:%02d %+03d%02d",
			tc->d.month, tc->d.day, tc->d.year,
			tc->t.hour, tc->t.minute, tc->t.second, tc->t.frame,
			tc->d.timezone/60, abs(tc->d.timezone)%60
			);
}

void timecode_time_to_string (TimecodeTime const * const t, char *smptestring) {
	snprintf(smptestring, 16, "%02d:%02d:%02d:%02d",
			t->hour, t->minute, t->second, t->frame);
}

/* C99 only defines strpbrk() - this is the reverse version */
static const char *strrpbrk(const char * const haystack, const char * const needle) {
	const char *ph, *pn, *rv;
	if (!haystack || !needle) return NULL;
	for (ph = haystack, rv = NULL ; *ph ; ++ph) {
		for (pn = needle; *pn != '\0' ; ++pn) {
			if (*pn == *ph) {
				rv = ph;
				break;
			}
		}
	}
	return rv;
}

void timecode_parse_time (TimecodeTime * const t, TimecodeRate const * const r, const char *val) {
	int i = 0;
	char *buf = strdup(val);
	char *pe;
	int32_t * const bcd[5] = {&t->frame, &t->second, &t->minute, &t->hour, NULL };

	t->subframe = 0;
	for (i=0; i<4; i++) {
		*bcd[i] = 0;
	}

	i=0;
	while (i < 4 && (pe= (char*) strrpbrk(buf,":.;"))) {
		*bcd[i] = (int) atoi(pe+1);
		*pe = '\0';
		i++;
	}

	if (i < 4 && *buf) {
		*bcd[i]= (int) atoi(buf);
	}

	free(buf);

	move_overflow(t, r);

	if (r->drop && (t->minute%10 != 0) && (t->second == 0) && (t->frame == 0)) {
		t->frame=2;
	}
}
